### PR TITLE
feat: bracket completion trigger

### PR DESCRIPTION
**Pull request by moliva**
This is an extension to #25.

Introduces `[` as a trigger character for completion, managing replacements avoiding to add wrapping characters if already present in file contents (mostly due to automatic edit tools completion).

### CODE REVIEW

**Comment by antonk52** [comment]
Thanks for the PR. I am currently traveling. I will try to review this over the weekend/next week when I get back.

File: src/CompletionProvider.ts [hunk 1]
```diff
@@ -11,20 +11,136 @@ import {
...
+    startsWithQuote: boolean;
+    endsWithQuote: boolean;
+};
+
+/**
+ * check if current character or last character is any of the completion triggers (i.e. `.`, `[`) and return it
+ *
+ * @see COMPLETION_TRIGGERS
+ */
+function findTrigger(line: string, position: Position): string | undefined {
```
Comment (antonk52) [inline src/CompletionProvider.ts:27]: nit: void should be preferred in return types if we do not intend to use it as a value
```suggestion
function findTrigger(line: string, position: Position): string | void {
```

File: src/CompletionProvider.ts [hunk 2]
```diff
@@ -11,20 +11,136 @@ import {
...
+    }
+
+    return undefined;
 }
 
-function getWords(line: string, position: Position): string {
+/**
+ * Given the line, position and trigger, returns the identifier referencing the styles spreadsheet and the (partial) field selected with options to help construct the completion item later.
+ *
+ * @returns
```
Comment (antonk52) [inline src/CompletionProvider.ts:45]: I assume this was not supposed to be commited

```suggestion
```

File: src/CompletionProvider.ts [hunk 3]
```diff
@@ -11,20 +11,136 @@ import {
...
+ */
+function getWords(
+    line: string,
+    position: Position,
+    trigger: string,
+): [string, string, FieldOptions?] | undefined {
```
Comment (antonk52) [inline src/CompletionProvider.ts:51]: nit: the same about return type
```suggestion
): [string, string, FieldOptions?] | void {
```

File: src/CompletionProvider.ts [hunk 4]
```diff
@@ -11,20 +11,136 @@ import {
...
+    if (words === '' || words.indexOf(trigger) === -1) {
+        return undefined;
+    }
+
+    switch (trigger) {
+        // process `.` trigger
+        case '.':
+            return words.split('.') as [string, string];
+        // process `[` trigger
+        default: {
```
Comment (antonk52) [inline src/CompletionProvider.ts:69]: Lets handle `[` as a case as well. The default case can throw an error as we do not expect other trigger characters

File: src/CompletionProvider.ts [hunk 5]
```diff
@@ -11,20 +11,136 @@ import {
 } from './utils';
 import type {CamelCaseValues} from './utils';
 
-// check if current character or last character is .
-function isTrigger(line: string, position: Position): boolean {
+export const COMPLETION_TRIGGERS = ['.', '[', '"', "'"] as const;
```
Comment (antonk52) [inline src/CompletionProvider.ts:14]: Do we want to handle backtick `\`` as well?

File: src/connection.ts [hunk 6]
```diff
@@ -42,9 +45,9 @@ export function createConnection(): lsp.Connection {
                 implementationProvider: true,
                 completionProvider: {
                     /**
-                     * only invoke completion once `.` is pressed
+                     * only invoke completion once `.` or `[` are pressed
                      */
-                    triggerCharacters: ['.'],
+                    triggerCharacters: COMPLETION_TRIGGERS.slice(),
```
Comment (antonk52) [inline src/connection.ts:50]: nit: I don't think we need to clone an array here

**Review by antonk52 — CHANGES_REQUESTED** [review]
Thanks for the wait

A few nits and overall looks great, I've tested this locally and it works as advertised. Happy to merge once the comments are addressed 🙂

File: src/CompletionProvider.ts (hunk 1 above)
Comment (moliva) [reply src/CompletionProvider.ts:27]: I do get the following lint error when applying this suggestion:
```
void is not valid as a constituent in a union type [lint/suspicious/noConfusingVoidType]
```
I think I could try and disable that rule, but wanted to double check your thoughts here.

File: src/CompletionProvider.ts (hunk 5 above)
Comment (moliva) [reply src/CompletionProvider.ts:14]: Adding both literal quotes (i.e. `'` and `"`) was a way to trigger the completion again as the editor would dismiss the completion options if you wrote the start of the string literal. I think we could add the backtick as well, as long as a literal class name is written (but not an interpolation expression) should still be handled, but would not provide extra value.

If it's ok with you, I would explore this as part of a following PR.

File: src/connection.ts (hunk 6 above)
Comment (moliva) [reply src/connection.ts]: This is the result of declaring this array as a `const`. I went ahead and removed it.

File: src/CompletionProvider.ts (hunk 1 above)
Comment (antonk52) [reply src/CompletionProvider.ts:27]: oh that's really interesting. I was not aware of this lint rule. Let's leave it as `undefined` then. From biome [docs](https://biomejs.dev/linter/rules/no-confusing-void-type/). it looks like they allow void as a standalone return type but not as union in return. Thanks for checking it out.

**Comment by antonk52** [comment]
Thanks!

Released in [1.4.0](https://github.com/antonk52/cssmodules-language-server/releases/tag/v1.4.0) and published to npm

**Pull request merged**